fix(date-picker): month and year picker scroller

buchananwill opened the pull request:
<!---
Thanks for creating a Pull Request ❤️!

Please read the following before submitting:
- PRs that adds new external dependencies might take a while to review.
- Keep your PR as small as possible.
- Limit your PR to one type (docs, feature, refactoring, ci, repo, or bugfix)
-->

Closes #3785

## 📝 Description

> Fixing the month/year UI sync bug.

## ⛳️ Current behavior (updates)

> The month and year were incorrectly updating due to a UI rectangle intersection error. This has been removed and the data is updated directly from the value attribute.

## 🚀 New behavior

> Please describe the behavior or changes this PR adds

## 💣 Is this a breaking change (Yes/No):

No.

## 📝 Additional Information


<!-- This is an auto-generated comment: release notes by coderabbit.ai -->
## Summary by CodeRabbit

## Summary by CodeRabbit

- **New Features**
	- Introduced the `AccordionTree` component for hierarchical data representation.
	- Added an aria label for the month/year picker toggle button to enhance accessibility.
	- Enhanced keyboard interaction capabilities in the calendar picker component.
	- Added a new custom hook for managing key press states and scroll events.
- **Bug Fixes**
	- Improved user interface and data synchronization for the month/year picker component.
- **Refactor**
	- Simplified scrolling behavior and focused date management in the calendar picker component.
- **Chores**
	- Updated versions for `@nextui-org/accordion` and `@nextui-org/calendar` packages.
	- Added a new test case for the month and year picker functionality in the `DatePicker` component.
<!-- end of auto-generated comment: release notes by coderabbit.ai -->

wingkwong reviewed (changes requested):
1. please include a before/after video and include steps on how to reproduce the issue.
2. please include the issue link if possible.
3. please remove accordion code. it's not related to this PR.
4. please include a test case.

wingkwong reviewed (changes requested):
just keep one changeset file where no accordion is mentioned

buchananwill commented:
> just keep one changeset file where no accordion is mentioned

Is that correct now?

buchananwill commented:
Please suggest an alternative way to mark this component so the test can
find it.

On Wed, 25 Sept 2024, 14:04 աӄա, ***@***.***> wrote:

> ***@***.**** requested changes on this pull request.
> ------------------------------
>
> In packages/components/calendar/src/calendar-header.tsx
> <https://github.com/nextui-org/nextui/pull/3789#discussion_r1775134096>:
>
> > @@ -99,6 +101,7 @@ export function CalendarHeader(props: CalendarHeaderProps) {
>    return showMonthAndYearPickers ? (
>      <Button
>        {...headerProps}
> +      aria-label={monthAndYearPickerToggle}
>
> we don't hardcode aria-label internally. We can take it from users' side
> tho.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/nextui-org/nextui/pull/3789#pullrequestreview-2328101810>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/BBL2M6NRIQAJMCWW6L6TKF3ZYKYEDAVCNFSM6AAAAABOU7COTKVHI2DSMVQWIX3LMV43YUDVNRWFEZLROVSXG5CSMV3GSZLXHMZDGMRYGEYDCOBRGA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

wingkwong commented:
In the test, you can simply get the element by `container.querySelector("[data-slot='header']")`

ryo-manba commented:
Please include the content of this comment in the PR description.
https://github.com/nextui-org/nextui/pull/3789#issuecomment-2367949681

ryo-manba commented on packages/components/date-picker/__tests__/date-picker.test.tsx:
```diff
@@ -600,6 +601,48 @@ describe("DatePicker", () => {
       expect(onHeaderExpandedChangeSpy).not.toHaveBeenCalled();
     });
 
+    it("should focus the month and year when pressed in the picker", () => {
```
I’m unclear about what you want to verify with this test. I think the issue is that the focus indicator misaligns when the up and down arrow keys are held down. We should verify that it is synchronized properly.

buchananwill replied:
Sorry, forgot to check the keyboard inputs as well. I've gone round the UI one more time, testing:

- mouse
- keyboard individual keystrokes
- held down keys

The current behaviour seems to be:
- mouse: blue focus and grey highlighter always in sync throughout interactions
- individual keystrokes: same
- HELD KEYS: the blue focus indicator always syncs to the focused date, but _sometimes_ it manages to travel beyond the grey highlighter, but will immediately return on the next **individual** keystroke.

This is from my perspective a big improvement, because the toggle on/off focus mutation is gone, and the only time the UI desyncs is _(sometimes)_ when you hold down a key.

ryo-manba reviewed (changes requested):
This implementation prevents switching values using the up and down arrow keys, so please fix that.

https://github.com/user-attachments/assets/07928b41-a308-42ec-92da-b4a92072835c

wingkwong commented:
@buchananwill please click re-request review once it's ready for review.

buchananwill commented:
I had one more idea, which I've added. The PickerItems now also have a keyUp listener, which just calls the scrollIntoView function. This seems to be a bullet proof way of ensuring that if you've been holding down any of the directional keys, then whatever was the last key to fire (and thus has focus) is scrolled to the desired position.

I have to confess my weaker experience with writing imperative DOM events, so please if anyone with that could ponder whether this broad-brush fix could have any unexpected side-effects?

I also changed the home/end values of the keyDown event to 1 and 12, since months are 1-indexed, not zero indexed. These buttons don't work for the year picker, since 1 and 12 aren't valid years within the available range. Couldn't see an obvious way to pass in the max/min years that would be dynamically set (since it could vary with locale etc.).

Hopefully this wraps up the functionality of this component & hook?

ryo-manba commented on packages/components/calendar/src/use-calendar-picker.ts:
```diff
@@ -221,11 +177,30 @@ export function useCalendarPicker(props: CalendarPickerProps) {
 
       const nextItem = map.get(nextValue);
 
+      scrollTo(nextValue, list);
```
Two things need fixing in the Home/End key behavior:
1. The year isn't changing.
2. It's jumping straight to the start/end instead of moving smoothly as it did before.

ryo-manba commented on packages/components/date-picker/__tests__/date-picker.test.tsx:
```diff
@@ -600,6 +601,48 @@ describe("DatePicker", () => {
+      expect(yearBefore).toHaveAttribute("tabindex", "-1");
+      triggerPress(monthBefore);
+      triggerPress(yearBefore);
+      expect(valueChangeSpy).not.toHaveBeenCalled();
+      expect(monthBefore).toHaveAttribute("tabindex", "0");
+      expect(yearBefore).toHaveAttribute("tabindex", "0");
```
Please also check the following two points:

1. Ensure that the value in the header is synchronized.
2. Verify that pressing the Enter key triggers the `valueChangeSpy`.

ryo-manba replied:
Please add tests.

ryo-manba commented:
The Home/End keys seem to already function in the [original implementation](https://nextui.org/docs/components/date-picker#with-month-and-year-pickers). Where did you find the information about the month's index?

> I also changed the home/end values of the keyDown event to 1 and 12, since months are 1-indexed, not zero indexed. These buttons don't work for the year picker, since 1 and 12 aren't valid years within the available range.

buchananwill commented:
I added two additional hooks to cover the different behaviours needed for the different navigation keys:

1. Home/End use a deferred focus that only triggers when the scroll ends. Otherwise, the focus event overrides the scrolling, and the element immediately snaps into view.
2. Home/End also use a repeated key blocker, to prevent the value from receiving more than one update from these keys, since 
    a. repetition of this event is meaningless
    b. it was playing havoc with the other scrolling/focus events/behaviours.

If you have existing hooks with the required API already in the codebase, feel free to swap them in. Equally, feel free to migrate them to a different (generic) hook folder if you want to centralize them for other scenarios to use.

I can't find any more bugs in this component.

On a side note, I prefer the simpler version where home/end cause the instant scrolling. I get that this is a minor inconsistency in the style of the UI, but the two extra control flow checks to defer the focus update _and_ block repeat key strokes seems overly convoluted. I wouldn't be surprised to find it has negative performance implications or introduces other bugs on different systems from my own...

Let me know what you think! :-)

ryo-manba commented:
> On a side note, I prefer the simpler version where home/end cause the instant scrolling. I get that this is a minor inconsistency in the style of the UI, but the two extra control flow checks to defer the focus update and block repeat key strokes seems overly convoluted. I wouldn't be surprised to find it has negative performance implications or introduces other bugs on different systems from my own...

This seems to be a separate issue from the current one. If you'd like to address it, please open a new issue for further discussion.

ryo-manba commented on packages/components/calendar/src/use-scroll-end-callback.ts:
```diff
@@ -0,0 +1,69 @@
+import {useCallback, useEffect, useRef} from "react";
+
+function useScrollEndCallback(debounce: number) {
```
The value is not being synced during mouse scroll. Please fix it.

buchananwill replied:
Done.

buchananwill replied:
Could you please specify the tests you need? Given that original component had tests that failed to catch this bug, it's a somewhat open-ended task. My current codebase uses with a different UI library, so it's difficult to justify spending any more time on the NextUI repo at the moment, but if you can ring-fence exactly what needs doing I will make time for that.

luka-on-gh commented:
Any updates on this PR?
